compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -85,13 +85,10 @@ impl<'hir> LoweringContext<'_, 'hir> {
                     hir::ExprKind::AddrOf(k, m, ohs)
                 }
                 ExprKind::Let(ref pat, ref scrutinee) => {
-                    self.lower_expr_let(e.span, pat, scrutinee)
+                    hir::ExprKind::Let(self.lower_pat(pat), self.lower_expr(scrutinee))
                 }
-                ExprKind::If(ref cond, ref then, ref else_opt) => match cond.kind {
-                    ExprKind::Let(ref pat, ref scrutinee) => {
-                        self.lower_expr_if_let(e.span, pat, scrutinee, then, else_opt.as_deref())
-                    }
-                    _ => self.lower_expr_if(cond, then, else_opt.as_deref()),
+                ExprKind::If(ref cond, ref then, ref else_opt) => {
+                    self.lower_expr_if(cond, then, else_opt.as_deref())
```

Comment:
If expressions will need to wrap anything that's not a `Let` expression in `DropTemps` and `compiler/rustc_passes/src/region.rs` will need updating to make the condition not be a terminating scope.